### PR TITLE
Bug 1780180: Backport metrics, console_url, metrics test [Release-4.1]

### PR DESCRIPTION
**Pull request by benjaminapetersen**
Need the 4.3 fix before I can bugzilla for this.

Enables `/metrics` endpoint with `console_url` & e2e tests in 4.1.

PR consists of:
- cherry pick the /metrics test
    - `git cherry-pick 9b38c92191` (with resolutions)
- cherry pick the metrics code:
    - git cherry-pick 346399b678

/cc @spadgett @jhadvig

### CODE REVIEW

**Comment by benjaminapetersen** [comment]
Closing for now.  Too much noise. WIll reopen backports when the 4.3 fix is settled.

**Comment by benjaminapetersen** [comment]
/retest

```

error: uploading the source layer sha256:7b1c937e0f6794db2535be6e4cb6d60a0b668ef78c2576611a3fb9c97a95ccdf failed: unable to upload layer sha256:7b1c937e0f6794db2535be6e4cb6d60a0b668ef78c2576611a3fb9c97a95ccdf to destination repository: received unexpected HTTP status: 504 Gateway Time-out
```

**Comment by benjaminapetersen** [comment]
/reteset

```
 level=fatal msg="failed to initialize the cluster: Cluster operator console has not yet reported success: timed out waiting for the condition" 
```
and

```

E1204 23:13:29.247853       1 reflector.go:134] github.com/openshift/client-go/operator/informers/externalversions/factory.go:101: Failed to list *v1.Console: consoles.operator.openshift.io is forbidden: User "system:serviceaccount:openshift-console-operator:console-operator" cannot list resource "consoles" in API group "operator.openshift.io" at the cluster scope
E1204 23:13:29.247979       1 reflector.go:134] github.com/openshift/client-go/config/informers/externalversions/factory.go:101: Failed to list *v1.Console: consoles.config.openshift.io is forbidden: User "system:serviceaccount:openshift-console-operator:console-operator" cannot list resource "consoles" in API group "config.openshift.io" at the cluster scope
E1204 23:13:29.265325       1 reflector.go:134] github.com/openshift/client-go/oauth/informers/externalversions/factory.go:101: Failed to list *v1.OAuthClient: oauthclients.oauth.openshift.io "console" is forbidden: User "system:serviceaccount:openshift-console-operator:console-operator" cannot list resource "oauthclients" in API group "oauth.openshift.io" at the cluster scope
E1204 23:13:29.266147       1 reflector.go:134] github.com/openshift/client-go/config/informers/externalversions/factory.go:101: Failed to list *v1.Infrastructure: infrastructures.config.openshift.io is forbidden: User "system:serviceaccount:openshift-console-operator:console-operator" cannot list resource "infrastructures" in API group "config.openshift.io" at the cluster scope
E1204 23:13:29.266330       1 reflector.go:134] github.com/openshift/client-go/config/informers/externalversions/factory.go:101: Failed to list *v1.ClusterOperator: clusteroperators.config.openshift.io is forbidden: User "system:serviceaccount:openshift-console-operator:console-operator" cannot list resource "clusteroperators" in API group "config.openshift.io" at the cluster scope
```
RBAC related errors, looking into this.

**Comment by benjaminapetersen** [comment]
/retest

Hmm. all of the above exist now in `manifests/03-rbac-role-cluster.yaml` and are unrelated to `/metrics` changes.

**Comment by benjaminapetersen** [comment]
```
FAIL	github.com/openshift/console-operator/test/e2e [setup failed]
```

**Comment by benjaminapetersen** [comment]
```
E1205 15:58:53.136008       1 streamwatcher.go:109] Unable to decode an event from the watch stream: http2: server sent GOAWAY and closed the connection; LastStreamID=181, ErrCode=NO_ERROR, debug=""
E1205 15:58:53.136305       1 streamwatcher.go:109] Unable to decode an event from the watch stream: http2: server sent GOAWAY and closed the connection; LastStreamID=181, ErrCode=NO_ERROR, debug=""
# ...
time="2019-12-05T15:58:55Z" level=info msg="deployment is available, ready replicas: 2 \n"
time="2019-12-05T15:58:55Z" level=info msg="sync_v400: updating console status"
time="2019-12-05T15:58:55Z" level=info msg="route ingress 'default' found and admitted, host: console-openshift-console.apps.ci-op-r767sk8i-bce1f.origin-ci-int-aws.dev.rhcloud.com \n"
```

**Comment by benjaminapetersen** [comment]
/retest

```
FAIL	github.com/openshift/console-operator/test/e2e [setup failed]
```
gonna go with possible flake unless it happens again.

**Comment by benjaminapetersen** [comment]
Added an additional PR backporting some helper funcs to ensure the test works.
I should probably reorder these to keep a neat history, the helpers before the actual test.  Lets see if prow complains.

**Comment by benjaminapetersen** [comment]
/retest

``` === RUN   TestMetricsEndpoint
--- FAIL: TestMetricsEndpoint (33.26s)
	metrics_test.go:29: waiting for setup to reach settled state...
	console-operator.go:361: waited 10 seconds to reach settled state...
	metrics_test.go:144: route for metrics: https://metrics-openshift-console-operator.apps.ci-op-2f9imn5j-bce1f.origin-ci-int-aws.dev.rhcloud.com/metrics
	metrics_test.go:150: route to /metrics: (https://metrics-openshift-console-operator.apps.ci-op-2f9imn5j-bce1f.origin-ci-int-aws.dev.rhcloud.com/metrics)
	metrics_test.go:57: fetching metrics url.... https://metrics-openshift-console-operator.apps.ci-op-2f9imn5j-bce1f.origin-ci-int-aws.dev.rhcloud.com/metrics
	metrics_test.go:59: fetching /metrics data...
	metrics_test.go:62: searching for console_url in metrics data...
	metrics_test.go:65: did not find console_url
	metrics_test.go:44: waiting for cleanup to reach settled state...
=== RUN   TestRemoved
--- FAIL: TestRemoved (300.11s)
	console-operator.go:153: changing console operator state to 'Removed'...
	logs.go:78: the latest observed state of the console resource:
		(*v1.Console)(0xc42001c200)({
		 TypeMeta: (v1.TypeMeta) &TypeMeta{Kind:,APIVersion:,},
		 ObjectMeta: (v1.ObjectMeta) &ObjectMeta{Name:cluster,GenerateName:,Namespace:,SelfLink:/apis/operator.openshift.io/v1/consoles/cluster,UID:cdd1a105-1850-11ea-beca-0a2405aed4b1,ResourceVersion:16562,Generation:9,CreationTimestamp:2019-12-06 17:49:51 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{release.openshift.io/create-only: true,},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},
		    }
		   }
		  }
		 }
		})
```

Can't tell if flake or what.  At this point, all tests fail after.  Looking through the logs.

File: pkg/console/transport/transport.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
+package transport
```
Comment (benjaminapetersen) [inline pkg/console/transport/transport.go]: This file is unnecessary, it should be removed.

/hold.

File: .gitignore [hunk 2]
```diff
@@ -9,6 +9,11 @@ _output
 tmp/_output
 tmp/_test
 
+# local dev scripts not intended for general use
```
Comment (benjaminapetersen) [inline .gitignore]: Consistency, avoid accidentally committing random scripts.

**Comment by spadgett** [comment]
4.1 will be at end of maintenance support when 4.4 ships. Given that this a not a serious bug, closing.

/close

**Pull request closed without merging**